### ISSUE TITLE
mps and cpu give far different results when training a transformer.

### ISSUE THREAD

**Issue by drmatthewclark** [pytorch/pytorch#103343]
### 🐛 Describe the bug

when training a transfromer model with mac CPU this is the resulting loss function log
994,tensor(13.4385),2.598125166893005,2.676572799682617
1988,tensor(13.2934),2.5872674012184143,2.6137495040893555
2000,tensor(13.3104),2.5885423374176026,2.628082513809204
2982,tensor(12.8851),2.5560683608055115,2.675567865371704
3976,tensor(12.6861),2.54050742149353,2.4734363555908203
4000,tensor(12.8042),2.549770863056183,2.584007501602173
4970,tensor(12.6474),2.537447814941406,2.528282403945923

when using the MPS, same code and data
994,tensor(1.4541),0.3743850642442703,0.42320576310157776
1988,tensor(1.4568),0.3762083804607391,0.3572617173194885
2000,tensor(1.4540),0.37433584868907926,0.3778816759586334
2982,tensor(1.4524),0.3732233664393425,0.3656735122203827
3976,tensor(1.4476),0.36990807741880416,0.3881590962409973
4000,tensor(1.4447),0.36788938045501707,0.3148590922355652
4970,tensor(1.4531),0.373729208111763,0.3972059488296509

[test.zip](https://github.com/pytorch/pytorch/files/11710031/test.zip)

### Versions

Collecting environment information...
PyTorch version: 2.1.0.dev20230608
Is debug build: False
CUDA used to build PyTorch: None
ROCM used to build PyTorch: N/A

OS: macOS 13.4 (arm64)
GCC version: Could not collect
Clang version: 14.0.3 (clang-1403.0.22.14.1)
CMake version: Could not collect
Libc version: N/A

Python version: 3.11.3 (main, Apr 19 2023, 18:49:55) [Clang 14.0.6 ] (64-bit runtime)
Python platform: macOS-13.4-arm64-arm-64bit
Is CUDA available: False
CUDA runtime version: No CUDA
CUDA_MODULE_LOADING set to: N/A
GPU models and configuration: No CUDA
Nvidia driver version: No CUDA
cuDNN version: No CUDA
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True

CPU:
Apple M2 Pro

Versions of relevant libraries:
[pip3] numpy==1.24.3
[pip3] torch==2.1.0.dev20230608
[pip3] torchaudio==2.1.0.dev20230608
[pip3] torchvision==0.16.0.dev20230608
[conda] numpy                     1.24.3          py311hb57d4eb_0  
[conda] numpy-base                1.24.3          py311h1d85a46_0  
[conda] pytorch                   2.1.0.dev20230608        py3.11_0    pytorch-nightly
[conda] torchaudio                2.1.0.dev20230608       py311_cpu    pytorch-nightly
[conda] torchvision               0.16.0.dev20230608       py311_cpu    pytorch-nightly

cc @kulinseth @albanD @malfet @DenisVieriu97 @razarmehr @abhudev

**Comment by drmatthewclark** [comment]
I abstracted the code in the zip file.   run "python demo.py train_seq eval_seq "  to reproduce the data shown.    
replace "cpu" with "mps" in the code between runs.

**Comment by ZhaoqiongZ** [comment]
Hi @drmatthewclark , I ran your code and data on cuda platform and the result is consistency with those on cpu, so I suppose it is a mps device issue.

**Comment by drmatthewclark** [comment]
> Hi @drmatthewclark , I ran your code and data on cuda platform and the result is consistency with those on cpu, so I suppose it is a mps device issue.

yes,  the mps code is totally not working correctly for training the  SimpleTransformers model.   I hope it is fixed soon.
I also ran it on an NVIDIA system to verify that the issue is with the  MPS implementation.

**Comment by drmatthewclark** [comment]
more insight -  If i create a model on NVIDIA, then open it on a Mac/mps system with map_location=torch.device('mps') then the loss values are close to that of the NVIDIA machine.   So there may be something related to initializing the model on the mps system.

**Comment by drmatthewclark** [comment]
this still behaves this way with current pytorch/transformers/ os 14.1.2

**Comment by watarungurunnn** [comment]
I commented out  `attn = at::dropout(attn, dropout_p, true);` in `pytorch/aten/src/ATen/native/transformers/attention.cpp`, line 729, and this passed the test.
It seems that the inconsistency is caused by the dropout function.
You can see that `nn.functional.dropout` is listed in `UNDEFINED_XFAILLIST`, which is consistent with my observation.

**Comment by watarungurunnn** [comment]
I found that the dropout function referred by attention module is causing the problem.
see: https://github.com/pytorch/pytorch/issues/103343#issuecomment-1851579023

**Comment by drmatthewclark** [comment]
the current build from source 16 Dec 2023 appears to have corrected this issue.   I assume that it will propogate to a release.
I did not check in changes but some updates from others discussing this topic, e.g. https://github.com/watarungurunnn 
identified here https://github.com/pytorch/pytorch/issues/111479  appear to have corrected the issue.

1 epoch MPS loss is 2.5,   1 epoch NVIDIA loss is 2.7.   because of the randomness involved in the training process these are comparable, where before they were different by closer to 10x
